src/main/cpp/src/get_json_object.hpp

```diff
@@ -0,0 +1,1191 @@
+      output_len += len;
+    }
+  }
+
+  /**
+   * write child raw value
+   * e.g.:
+   *
+   * write_array_tokens = false
+   * need_comma = true
+   * [1,2,3]1,2,3
+   *        ^
+   *        |
+   *    child pointer
+   * ==>>
+   * [1,2,3],1,2,3
+   *
+   *
+   * write_array_tokens = true
+   * need_comma = true
+   *   [12,3,4
+   *     ^
+   *     |
+   * child pointer
+   * ==>>
+   *   [1,[2,3,4]
+   *
+   * @param child_block_begin
+   * @param child_block_len
+   */
+  __device__ void write_child_raw_value(char* child_block_begin,
+                                        size_t child_block_len,
+                                        bool write_outer_array_tokens)
+  {
+    bool insert_comma = need_comma();
+
+    is_first_item[array_depth - 1] = false;
+
+    if (nullptr != output) {
+      if (write_outer_array_tokens) {
+        if (insert_comma) {
+          *(child_block_begin + child_block_len + 2) = ']';
+          move_forward(child_block_begin, child_block_len, 2);
+          *(child_block_begin + 1) = '[';
+          *(child_block_begin)     = ',';
+        } else {
+          *(child_block_begin + child_block_len + 1) = ']';
+          move_forward(child_block_begin, child_block_len, 1);
+          *(child_block_begin) = '[';
+        }
+      } else {
+        if (insert_comma) {
+          move_forward(child_block_begin, child_block_len, 1);
+          *(child_block_begin) = ',';
+        } else {
+          // do not need comma && do not need write outer array tokens
+          // do nothing, because child generator buff is directly after the
+          // parent generator
+        }
+      }
+    }
+
+    // update length
+    if (insert_comma) { output_len++; }
+    if (write_outer_array_tokens) { output_len += 2; }
+    output_len += child_block_len;
+  }
+
+  __device__ void move_forward(char* begin, size_t len, int forward)
+  {
+    char* pos = begin + len + forward - 1;
+    char* e   = begin + forward - 1;
+    // should add outer array tokens
+    // First move chars [2, end) a byte forward
+    while (pos > e) {
+      *pos = *(pos - 1);
+      pos--;
+    }
+  }
+
+  __device__ void reset() { output_len = 0; }
+
+  __device__ inline size_t get_output_len() const { return output_len; }
+  __device__ inline char* get_output_start_position() const { return output; }
+  __device__ inline char* get_current_output_position() const { return output + output_len; }
+
+  /**
+   * generator may contain trash output, e.g.: generator writes some output,
+   * then JSON format is invalid, the previous output becomes trash.
+   */
+  __device__ inline void set_output_len_zero() { output_len = 0; }
+
+  __device__ inline void set_output_len(size_t len) { output_len = len; }
+
+ private:
+  char* output;
+  size_t output_len;
+  bool hide_outer_array_tokens;
+
+  bool is_first_item[max_json_nesting_depth];
+  int array_depth = 0;
+};
+
+/**
+ * path evaluator which can run on both CPU and GPU
+ */
+struct path_evaluator {
```

Comment:
Nit: Device code and/or class with device code should be placed in `.cu` or `.cuh` files. `.hpp` files should never contain device code.